internal/driver/ldap/main.go

```diff
@@ -0,0 +1,112 @@
+package ldap
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"regexp"
+	"text/template"
+
+	client "github.com/go-ldap/ldap/v3"
+	"github.com/peeley/carpal/internal/config"
+	"github.com/peeley/carpal/internal/driver"
+	"github.com/peeley/carpal/internal/resource"
+	"gopkg.in/yaml.v3"
+)
+
+type LdapClient interface {
+	Bind(string, string) error
+	Close() error
+	Search(*client.SearchRequest) (*client.SearchResult, error)
+}
+
+type ldapDriver struct {
+	Configuration config.Configuration
+	Template      *template.Template
+	ClientFunc    func() (LdapClient, error)
+}
+
+func NewLDAPDriver(conf config.Configuration) driver.Driver {
+	d := ldapDriver{
+		Configuration: conf,
+	}
+	d.Template = template.Must(template.ParseFiles(conf.LDAPConfiguration.Template))
+	d.ClientFunc = func() (LdapClient, error) {
+		return client.DialURL(conf.LDAPConfiguration.URL)
+	}
+	return d
+}
+
+func (d ldapDriver) GetResource(name string) (*resource.Resource, error) {
+	var resource resource.Resource
+
+	re, err := regexp.Compile("(.+):([^@]+)")
+	if err != nil {
+		return nil, err
+	}
+	resourceSplit := re.FindStringSubmatch(name)
+	if len(resourceSplit) == 0 {
+		return nil, driver.ResourceNotFound{ResourceName: name}
+	}
+	resourcename := resourceSplit[1:]
+
+	if len(resourcename) < 2 {
+		return nil, errors.New("Error breaking down resource")
+	}
+	if resourcename[0] != "acct" {
+		return nil, driver.ResourceNotFound{ResourceName: name}
+	}
+
+	username := resourcename[1]
+	c, err := d.ClientFunc()
+	if err != nil {
+		return nil, err
+	}
+	defer c.Close()
+	err = c.Bind(d.Configuration.LDAPConfiguration.BindUser, d.Configuration.LDAPConfiguration.BindPass)
+	if err != nil {
+		return nil, err
+	}
+
+	searchString := fmt.Sprintf("(%s=%s)", d.Configuration.LDAPConfiguration.UserAttr, username)
+	if d.Configuration.LDAPConfiguration.Filter != "" {
+		searchString = fmt.Sprintf("(&%v%v)", d.Configuration.LDAPConfiguration.Filter, searchString)
```

Comment:
While testing, I noticed these errors are propagated to the text of the HTTP response. This isn't ideal, since we probably don't want to leak LDAP configuration details to arbitrary clients. A little beyond the scope of this PR, but I can edit the error handling code to just return a generic `502 Bad Gateway` response or something.